File: solutions/Golang/populating_next_right_pointers_ii_queue.go
package main

import "fmt"

/**
 * Definition for a Node.
 */
type Node struct {
    Val   int
    Left  *Node
    Right *Node
    Next  *Node
}

/**
 * Populates next pointers using level-order BFS queue.
 * Time: O(n), Space: O(w) where w is max width
 */
func connect(root *Node) *Node {
    if root == nil {
        return root
    }

    queue := []*Node{root}

    for len(queue) > 0 {
        levelSize := len(queue)
        var prev *Node

        for i := 0; i < levelSize; i++ {
            node := queue[0]
            queue = queue[1:]

            if prev != nil {
                prev.Next = node
            }
            prev = node

            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }
    }

    return root
}

func main() {
    fmt.Println("Example 1: Tree with next pointers connected via queue")
}

File: solutions/Python/populating_next_right_pointers_ii_queue.py
from typing import Optional
from collections import deque


# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


def connect(root: Optional[Node]) -> Optional[Node]:
    """
    Populates next pointers in a perfect binary tree using level-order BFS queue.
    Time: O(n), Space: O(w) where w is max width
    """
    if not root:
        return root

    queue = deque([root])

    while queue:
        level_size = len(queue)
        prev = None

        for i in range(level_size):
            node = queue.popleft()

            if prev:
                prev.next = node
            prev = node

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

    return root


# Test cases
if __name__ == "__main__":
    # Example 1: Perfect binary tree
    root = Node(1,
                Node(2, Node(4), Node(5)),
                Node(3, Node(6), Node(7)))
    result = connect(root)
    print("Example 1: Tree with next pointers connected via queue")

    # Example 2: Single node
    single = Node(1)
    result = connect(single)
    print("Example 2: Single node tree")

    # Example 3: None
    result = connect(None)
    print("Example 3: None input")
